### ISSUE TITLE
MoonScript `self` references are counted as "malformed"

### ISSUE THREAD

**Issue by RyanSquared** [lunarmodules/LDoc#255]
Malformed reference when trying to use `\` in a reference in a MoonScript class. Patch below.

```diff
diff --git a/ldoc/doc.lua b/ldoc/doc.lua
index e19b2df..7015482 100644
--- a/ldoc/doc.lua
+++ b/ldoc/doc.lua
@@ -1098,7 +1098,7 @@ function Module:process_see_reference (s,modules,istype)
    local mod_ref,fun_ref,name,packmod
    local ref = custom_see_references(s)
    if ref then return ref end
-   if not s:match '^[%w_%.%:%-]+$' or not s:match '[%w_]$' then
+   if not s:match '^[%w_%.\\%:%-]+$' or not s:match '[%w_]$' then
       return nil, "malformed see reference: '"..s..'"'
    end

```

How to apply diff:

```sh
git apply << EOF
diff --git a/ldoc/doc.lua b/ldoc/doc.lua
index e19b2df..7015482 100644
--- a/ldoc/doc.lua
+++ b/ldoc/doc.lua
@@ -1098,7 +1098,7 @@ function Module:process_see_reference (s,modules,istype)
    local mod_ref,fun_ref,name,packmod
    local ref = custom_see_references(s)
    if ref then return ref end
-   if not s:match '^[%w_%.%:%-]+$' or not s:match '[%w_]$' then
+   if not s:match '^[%w_%.\\\\%:%-]+$' or not s:match '[%w_]$' then
       return nil, "malformed see reference: '"..s..'"'
    end

EOF
```

**Comment by ChickenNuggers** [comment]
issue also found with @name in @function | @lfunction, patching now

**Comment by nonchip** [comment]
hey @stevedonovan, maybe you want to fix this? it's a half year old issue even with a nice working patch and IT PREVENTS USING ANY REFERENCES IN MOONSCRIPT! -_-

**Comment by stevedonovan** [comment]
Man, I didn't realize it was that bad!  @RyanSquared, would you like to prepare a PR or shall I just do this manually? Useful to have a test case example.

**Comment by nonchip** [comment]
test case:
```moonscript
--- @module a

--- is a class
-- @type B
class B
  --- does stuff
  c: => print 'stuff'
  --- does other stuff by calling @{B\c}
  d: => @c!
  --- does even more stuff by calling @{d}
  e: => @d!
```
* names like `B\c` are generated but `@{B\c}` can't be resolved because `\` is "malformed"
* `@{B.c}` and `@{B:c}` can't be resolved because they're not generated
* `@{d}` also can't be resolved, but that's because @type/etc are not module-style containers, that's fine though if you know it.

**Comment by nonchip** [comment]
also, kudos to your config file sandboxing, i just tried to wrap the function in there to provide a temporary fix, and it din't let me :P

though you might want to consider allowing `require` there for more sophisticated callbacks like `custom_display_name_handler` etc.

**Comment by RyanSquared** [comment]
So I've come up with another issue; while my above patch does allow for referencing classes in the form of `-- @see x\y`, classes must be given their own file. @stevedonovan - if you want, merge my patch yourself. As for the classes, I'm not sure whether I'm happy with my solution. It looks like I'm going to have to settle with `-- @classmod Signer` if I want the class name, or `-- @module bassoon` for what to `require()`.

**Comment by nonchip** [comment]
@RyanSquared you can. classes inside modules are called `@type`.

**Comment by RyanSquared** [comment]
Do you mind providing an example?

**Comment by nonchip** [comment]
sure:
```moonscript
--- @module some.module
--- @type SomeType
class SomeType
```

**Comment by nonchip** [comment]
for a practical example just take a look at: https://github.com/cuddlyrobot/dysnomia/blob/master/modules/dysnomia/routers/regex_router.moon (using custom config https://github.com/cuddlyrobot/dysnomia/blob/master/config.ld and rather hackily applying your patch in https://github.com/cuddlyrobot/dysnomia/blob/master/setup.zsh#L68)

**Comment by RyanSquared** [comment]
So, I found out what the problem was - in my tests, the `new: =>` functions were not detected. I tried `-- @type` and it didn't show up. Turns out, after testing, only the `new` function did this? `_:` worked fine, as did other function names.

**Comment by nonchip** [comment]
yeah apparently LDoc thinks a class can't be a class if it doesn't have a constructor. which is strange since it doesn't even handle the constructor differently (note how it claims there's a function `MyClass\new` even though you instantiate it by calling `MyClass!`)

**Comment by RyanSquared** [comment]
I think it's because Lua's constructors are `__call`?

**Comment by nonchip** [comment]
yeah but it could be just documented as a function `MyClass(the_arguments_from_new)` because that's how you'd use it. and usually documentation tells you how to use something, not how it was written ;)

**Comment by nonchip** [comment]
but i'd guess most people using moonscript will know what to do.

**Comment by alerque** [comment]
This is fixed by 9fa1150 merged from #273.
